### ISSUE TITLE
Understanding the Double Discounting in the REINFORCE and VPG Algorithms

### ISSUE THREAD

**Issue by ZachariahRosenberg** [mimoralea/gdrl#9]
I noticed a double discounting in both the REINFORCE algorithm and VPG that is not in the original REINFORCE/VPG papers:

```
def optimize_model(self):
    T = len(self.rewards)
    discounts = np.logspace(0, T, num=T, base=self.gamma, endpoint=False)

    # returns are the discounted cumulative sum of rewards in the trajectory
    returns = np.array([np.sum(discounts[:T-t] * self.rewards[t:]) for t in range(T)])

   #< ... >
    # the loss here multiplies the discounted cumulative sum return by discounts again
    policy_loss = -(discounts * returns * self.logpas).mean()
```

In the original REINFORCE algorithm paper, the algorithm is described as:

![reinforce](https://user-images.githubusercontent.com/6923376/111916487-8a2b2700-8a51-11eb-932a-b43d5e89021c.png)

Where the loss is simply the negative mean `returns * log_probs`.

---
The author explains in his book, "Grokking DRL", 

> Notice here that we're using the mathematically correct policy-gradient update, which isn't what you commonly find out there. The extra discount assumes that we're trying to optimize the expect discounted return [sic] from the initial state, so returns later in an episode get discounted.

I'm not sure I understand how discounting a second time is what sets the target to the expected discounted return - I'd have thought that we are already trying to optimize for the expected discounted return by discounting the return to begin with.

Is there a paper or article on re-discounting the parameter update?

Thank you so much! Also, I'm loving the book!

**Comment by mimoralea** [comment]
Sutton's book is a good one: http://www.incompleteideas.net/book/the-book-2nd.html, see page 350 of the PDF.

Also, his policy gradient paper may help: https://papers.nips.cc/paper/1999/hash/464d828b85b0bed98e80ade0a5c43b0f-Abstract.html. Though, I recommend the book.
